src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -42,23 +43,59 @@ class AsyncLogWriter::AsyncLogLocker : public StackObj {
   }
 };
 
+// Reserve space for a flush token, so 'push_back(Token)' always succeeds.
+AsyncLogWriter::Buffer::Buffer(char* buffer, size_t capacity) : _pos(0), _buf(buffer) {
+  assert(capacity >= (AsyncLogWriter::Token.size()), "AsyncLogBuffer is too small!");
+  _capacity = capacity - (Token.size());
+}
+
+bool AsyncLogWriter::Buffer::push_back(const AsyncLogMessage& msg) {
+  size_t sz = msg.size();
+
+  if (_pos + sz <= _capacity || (&msg == &AsyncLogWriter::Token)) {
+    memcpy(_buf + _pos, &msg, sizeof(AsyncLogMessage));
+    char* dest = _buf + _pos + sizeof(AsyncLogMessage);
+    if (msg.message() != nullptr) {
+      strcpy(dest, msg.message());
+    } else {
+      dest = nullptr;
+    }
+    auto p = reinterpret_cast<AsyncLogMessage*>(_buf + _pos);
+    p->set_message(dest);
+    _pos += sz;
+    return true;
+  }
+
+  return false;
+}
+
+AsyncLogMessage* AsyncLogWriter::Buffer::Iterator::next() {
+  auto msg = static_cast<AsyncLogMessage*>(raw_ptr());
+  _curr += msg->size();
+  _curr = MIN2(_curr, _buf._pos);
+  return msg;
+}
+
+const LogDecorations& AsyncLogWriter::None = LogDecorations(LogLevel::Warning, LogTagSetMapping<LogTag::__NO_TAG>::tagset(),
+                                      LogDecorators::None);
+// Note: there is an initializer order issue here. C++ cannot guarantee that 'None' is initialized before Token!
```

Comment:
Sorry, this comment was misleading. I left it as a reminder. I mean c++ can't guarantee to initialize None and Token before `LogDecorators::None` is initialized. Like you said, LogDecorators::None is defined in another compilation unit. 

The reason it's right because `LogDecorators::None` is expected to be zero anyway. It's in bss section. That's why I feel it's tricky. I will change this part!